### ISSUE TITLE
x/tools/gopls: initial diagnostics not sent for file

### ISSUE THREAD

**Issue by myitcv** [golang/go#36144]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version devel +da4d58587e Sat Dec 7 15:57:30 2019 +0000 linux/amd64
$ go list -m golang.org/x/tools
golang.org/x/tools v0.0.0-20191213221258-04c2e8eff935 => github.com/myitcvforks/tools v0.0.0-20191214162001-758e20ac67bf
$ go list -m golang.org/x/tools/gopls
golang.org/x/tools/gopls v0.1.8-0.20191213221258-04c2e8eff935 => github.com/myitcvforks/tools/gopls v0.0.0-20191214162001-758e20ac67bf
</pre>

Note this version is a temporary fork with fixes as described in https://github.com/golang/go/issues/35638#issuecomment-565718553

### Does this issue reproduce with the latest release?

Yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE="on"
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/myitcv/.cache/go-build"
GOENV="/home/myitcv/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/myitcv/gostuff"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/home/myitcv/gos"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/home/myitcv/gos/pkg/tool/linux_amd64"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/myitcv/gostuff/src/github.com/myitcv/govim/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build705453102=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

### What did you do?

We have a [`govim` test](https://github.com/govim/govim/blob/44bd381ebb2db0be3a2ed1c9d58c9553308c3055/cmd/govim/testdata/scenario_default/gofmt.txt) that starts by opening a file and then calling `Formatting`.

But before calling `Formatting` we first wait for the initial diagnostics to be sent for the file.

However very occasionally we do not see the initial diagnostics for the file and the test times out (and fails) waiting.

Here is a failing and passing log of this test:

* [fail.log](https://github.com/golang/go/files/3964231/fail.log)
* [pass.log](https://github.com/golang/go/files/3964232/pass.log)

On the occasions that this test fails, changes to `go.mod` appear to be related. `govim` currently supports file watching by simulating the file that has changed being open in the editor. For example, with `go.mod` changes, `govim` detects the file has changed, signals `DidOpen` to `gopls` on the first change, then `DidChange` thereafter. 

As you can see from the failing log above, the calculation of diagnostics for `file.go` gets almost to the point of publishing, but then its context gets cancelled. That in itself isn't a problem except that the diagnostics are not subsequently sent. 

### What did you expect to see?

Initial diagnostics to be sent for a file.

### What did you see instead?

As above.

---

cc @stamblerre @heschik @ridersofrohan

**Comment by heschik** [comment]
This seems like a fairly deep-rooted problem. As I understand it, if we're diagnosing package A, and I edit a file in unrelated package B, we're going to cancel the first diagnostics run and then not re-diagnose A. The fact that it was a `go.mod` file this time may be something of a red herring.

**Comment by myitcv** [comment]
FWIW, this is now the main cause of test failures in the `govim` test suite. 

cc @findleyr

**Comment by myitcv** [comment]
@stamblerre I'm minded to close this given the massive number of updates recently, unless you/@heschik would like to keep it open?

**Comment by myitcv** [comment]
FWIW I'm re-enabling a test against current master to stress this out a bit: https://github.com/govim/govim/pull/684

**Comment by stamblerre** [comment]
Yes, I think it's reasonable to close this now that we diagnose entire workspace on every file change. Even if one diagnostics run gets canceled, another one will immediately be started and will complete once the user stops typing.

**Comment by myitcv** [comment]
Given the test case originally described in this issue, I think https://github.com/golang/go/issues/36463 is actually more relevant:

> The initial workspace load must not be canceled

Once we get a stable master (following a conclusion of https://github.com/golang/go/issues/36601) we can easily reenable our test that verifies whether things are working/not here.

**Comment by stamblerre** [comment]
Sounds good. Please follow up here once you get a chance to test this.
